### DISCUSSION TITLE
Function calling

### DISCUSSION THREAD

**Discussion by gruckion** [open-webui/open-webui#1629] [Ideas]
# Function calling

I will be compiling together research notes and feature scope for the `function calling` feature on the roadmap.

Feel free to reach out to provide feedback, suggestions on existing overlapping projects and ideas for technical solutions.

## Related issues

https://github.com/open-webui/open-webui/issues/851
https://github.com/open-webui/open-webui/issues/1321

## What is it? 

> 🔄 Function Calling: Empower your interactions by running code directly within the chat. Execute functions and commands effortlessly, enhancing the functionality of your conversations.

LLMs are powerful but they need to be able to execute code to get answer to users questions.

## Features

I will use the [MoSCoW method ](https://en.wikipedia.org/wiki/MoSCoW_method)to classify the features.

### Must have
- Allow AI and users to directly execute code for problem solving.
- Executed code should run inside a sandbox environment and not compromise the host.
- Allow resolving third party dependencies required by the generated code.
- Allow the executed code to save content outputs.
- Allow the outputted content to be rendered visually.
- Model agnostic, should be capable of executing code generated by any model.
 
### Could have
- Ability to access and control the local computer.

## Technical specification

@A-daily proposed leveraging [open-interpreter](https://github.com/KillianLucas/open-interpreter). This has 47.5k stars at the time of writing this, 104 open issues, 718 closed, 100 contributors and is 99.2% python. The tool is a command line based application and python package installed and is easy to get up and running. However, this tool executes code on your own local machine. Therefore it is not a desired solution but they do have plans to expand to using Docker on their [roadmap](https://github.com/OpenInterpreter/open-interpreter/blob/main/docs/ROADMAP.md).

A basic prototype has been provided by @zabirauf [here](https://github.com/zabirauf/open-webui/pull/1/files). Feedback on this prototype has requested the ability for the code execution to take place inside a safe isolated sandbox environment and should not compromise the host.

@zabirauf went on to suggest the use of [Piston](https://github.com/engineer-man/piston) which allows for executing code in an isolated manner updated prototype can be found [here](https://github.com/zabirauf/open-webui/blob/8dff4e734113aa6fcbd70a5ec35bd689910350f6/backend/apps/code/main.py). At the time of writing this it has 1.7k stars, 72 open issues and  216 closed. Languages JS:79.9%. It has a wide range of [supported languages](https://github.com/engineer-man/piston?tab=readme-ov-file#supported-languages). Makes use of Docker for sandboxing, the code is written to a temporary directory and run or compiled/run. Finally some effort has been made into security considerations which can be found [here](https://github.com/engineer-man/piston?tab=readme-ov-file#security).

1. Use subprocess to execute python code on the host.
    - PRO: easy to implement
    - CON: insecure contaminants the host.
2. Use docker python library to execute via docker containers
    - PRO: 
    - CON: 
3. Docker in docker (DIND)
    - PRO: 
    - CON: 

## References
https://github.com/OpenInterpreter/open-interpreter
https://platform.openai.com/docs/assistants/tools/code-interpreter/code-interpreter-beta
https://github.com/zabirauf/open-webui/pull/1/files
https://github.com/engineer-man/piston

**Comment by santosh0787** [comment]
hello, https://github.com/iamgreggarcia/codesherpa this also looks good.

**Comment by frederikschubert** [comment]
Also relevant https://github.com/e2b-dev/code-interpreter

**Comment by not-nullptr** [comment]
i've adapted my [personal webui](https://github.com/not-nullptr/ollama-functions)'s function calling code in PR #2175. the code for the functions themselves is written in monaco (vscode's editor), inside of the user's browser, using typescript. the code runs in the user's browser, too. this adds ease for quickly writing a function, but also means interfacing with your local PC is harder. suggestions welcome!!

**Comment by EtiennePerot** [comment]
Hi all. I work on [gVisor](https://gvisor.dev), the open-source sandboxing technology [used by ChatGPT for code execution](https://drive.google.com/file/d/1jjqrV76-86rdEcmFNnxMs4lI-ncAookn/view), as [mentioned in their security infrastructure blog post](https://openai.com/index/securing-research-infrastructure-for-advanced-ai/). gVisor is also used by Google as a sandbox when running user-uploaded code, such as in [Cloud Run](https://cloud.google.com/run/docs/securing/security).

Unlike previously-mentioned solutions, gVisor does not have external server dependencies, does not require virtualization (VM) support, and does not expose the host kernel the way Docker (or hardened configurations of Docker containers) do.

If anyone is interested in using it, I have recently implemented an Open WebUI function and tool for Python and Bash code execution that uses it. It also supports time-limited and RAM-limited execution.

| **Code execution function** | **Code execution tool** |
| --------------------------- | ----------------------- |
| ![Code execution function](https://github.com/EtiennePerot/open-webui-code-execution/blob/master/res/code-execution-function.gif?raw=true) | ![Code execution tool](https://github.com/EtiennePerot/open-webui-code-execution/blob/master/res/code-execution-tool.gif?raw=true) |
| Click button to run code block. | Grant the LLM the ability to run code by itself. |

You can [install them here](https://github.com/EtiennePerot/open-webui-code-execution).

**Reply by olibols** [reply]
Hi Mr Etienne. After toiling about for a while trying to get this work I get the following error from the LLM:
{"status": "ERROR", "output": "running container: creating container: cannot create sandbox: cannot read client sync file: waiting for sandbox to start: EOF"}

Any ideas?

**Reply by EtiennePerot** [reply]
Thanks for trying it out. Unfortunately, this error message alone isn't really enough to figure out what's going wrong.

I uploaded a new version of the tool to [this repository](https://github.com/EtiennePerot/open-webui-code-execution) that includes more debug info. I also made it so that you can run the script in debug mode outside of OpenWebUI, by just running it as a regular Python script. That makes it easier to debug in isolation. So try `docker exec -it container_name_here bash` to enter the OpenWebUI container you're running, then run:

```
echo 'import datetime; print(datetime.datetime.now())' | python3 /data/tools/run_code.py --debug
```

This will simulate what happens when the LLM tries to execute the Python code `import datetime; print(datetime.datetime.now())` in the sandbox. If it works well, you'll get output like this:

```
$ echo 'import datetime; print(datetime.datetime.now())' | python3 /data/tools/run_code.py --debug
{'type': 'status', 'data': {'status': 'in_progress', 'description': 'Checking if environment supports sandboxing...', 'done': False}}
{'type': 'status', 'data': {'status': 'in_progress', 'description': 'Initializing sandbox configuration...', 'done': False}}
{'type': 'status', 'data': {'status': 'in_progress', 'description': 'Setting up sandbox environment...', 'done': False}}
{'type': 'status', 'data': {'status': 'in_progress', 'description': 'Running Python code in gVisor sandbox...', 'done': False}}
{'type': 'status', 'data': {'status': 'complete', 'description': 'Python code executed successfully.', 'done': True}}
[... a whole bunch of debug information ...]
{"status": "OK", "output": "2024-08-30 06:47:42.473391"}
```

I expect that this won't actually work for you, and that instead you'll get the same error message at the end (`cannot read client sync file: waiting for sandbox to start: EOF`). However this time there will be useful debug output in there. Please post all of that as [an issue report on the tool's repository](https://github.com/EtiennePerot/open-webui-code-execution) rather than here, as this discussion is about OpenWebUI's function calling functionality in general, not really the best place for support on this specific tool.
